### ISSUE TITLE
Potential .NET Core support via System.Composition

### ISSUE THREAD

**Issue by tillig** [autofac/Autofac.Mef#9]
Based on discussions on [this issue](https://github.com/autofac/Documentation/issues/38) it appears there are "System.Composition" packages out that have some classic MEF (not MEF 2) functionality.

Would these allow us to support .NET Core with this package?

**Comment by matthew-a-thomas** [comment]
I have an example of using MEF to autodiscover Autofac `IModule`s, then registering them using Autofac's `RegisterModule` method. This is .Net Core 1.1. It seems to be working okay so far.

Check out the `RegisterModules` method here:
https://github.com/matthew-a-thomas/gps-logger/blob/4393570209a8b2a27d5ce9295abe86e9e2eb1ce8/GPSLogger/Startup.cs

Let me know if anyone has any questions!

Disclaimer: I admit a bit of [cargo cult programming](https://en.wikipedia.org/wiki/Cargo_cult_programming)  because I don't completely understand everything about MEF yet in [Ricardo Peres's example here](https://weblogs.asp.net/ricardoperes/using-mef-in-net-core). So perhaps my example could be cleaner--I'm open to suggestions!

**Comment by tillig** [comment]
I don't think you need MEF for any of that. Just loading a set of assemblies and finding types that implement a particular interface is pretty standard reflection.

The big point of this integration is to try getting `[Exports]` and `[Imports]` attributes to get into Autofac; or to create MEF based strongly typed metadata attributes. If you're not doing either of those things (doesn't look like it in your example) you don't need this integration library.

**Comment by matthew-a-thomas** [comment]
@tillig Ahh, I thought that since I had decorated [`SQLDatabase.Module`](https://github.com/matthew-a-thomas/gps-logger/blob/4393570209a8b2a27d5ce9295abe86e9e2eb1ce8/SQLDatabase/Module.cs) with the `Export` attribute, *that's* what was doing the trick. But you're right about that not really making a difference in my example, because [`GPSLogger.CompositionRoot`](https://github.com/matthew-a-thomas/gps-logger/blob/4393570209a8b2a27d5ce9295abe86e9e2eb1ce8/GPSLogger/CompositionRoot.cs) is not decorated with an attribute and yet is picked up also.

I'll try to come up with a different example that doesn't involve preloading assemblies.

**Comment by bencyoung** [comment]
I'd also really love a MEF2 .NET Core supporting Autofac.MEF. I've been having a play and it seems possible, but I'd miss out a lot of the subtleties that are supported here

**Comment by bencyoung** [comment]
Here's some code that does something towards what we want. I'm happy for other people to use it:

```csharp
using System;
using System.Collections.Generic;
using System.Composition;
using System.Composition.Hosting;
using System.Composition.Hosting.Core;
using System.Reflection;

using Autofac.Core;
using Autofac.Core.Registration;
using Autofac.Core.Activators.ProvidedInstance;
using Autofac.Core.Lifetime;
using Autofac;
using System.Linq;
using Autofac.Builder;

namespace DITest
{
    class Program
    {
        static void Main(string[] args)
        {
            var mefContainerBuilder = new ContainerConfiguration()
                .WithAssembly(Assembly.GetExecutingAssembly());

            var mefContainer = mefContainerBuilder.CreateContainer();

            var autoContainerBuilder = new ContainerBuilder();
            var source = new Source();
            var provider = new ExportProvider();
            autoContainerBuilder.Properties["MefContainerProvider"] = provider;
            autoContainerBuilder.RegisterSource(source);
            autoContainerBuilder.RegisterBuildCallback(cont =>
            {
                provider.Container = cont;
                mefContainerBuilder = mefContainerBuilder.WithProvider(provider);
                source.CompositionContainer = mefContainerBuilder.CreateContainer();
            });

            var exports = mefContainer.GetExports<IPluginRegistration>();
            foreach (var export in exports)
            {
                export.Register(autoContainerBuilder);
            }

            var container = autoContainerBuilder.Build();

            Console.WriteLine($"Found {container.Resolve<ITest>()} exported");
        }
    }

    public static class RegistrationExtensions
    {
        public static IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> Exported<TLimit, TActivatorData, TSingleRegistrationStyle>(this IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> registration)
            where TSingleRegistrationStyle : SingleRegistrationStyle
        {
            return registration.OnRegistered(reg =>
            {
                var mefProvider = (ExportProvider)reg.ComponentRegistry.Properties["MefContainerProvider"];
                foreach (var service in reg.ComponentRegistration.Services.OfType<IServiceWithType>())
                {
                    mefProvider.Registrations.Add(service.ServiceType, cont => cont.Resolve(service.ServiceType));
                }
            });;
        }
    }

    public class Source : IRegistrationSource
    {
        public CompositionHost CompositionContainer { get; set; }

        bool IRegistrationSource.IsAdapterForIndividualComponents => false;

        IEnumerable<IComponentRegistration> IRegistrationSource.RegistrationsFor(Service service, Func<Service, IEnumerable<IComponentRegistration>> registrationAccessor)
        {
            if (service is IServiceWithType typed && this.CompositionContainer != null)
            {
                if (this.CompositionContainer.TryGetExport(typed.ServiceType, out object exported))
                {
                    return new[]
                    {
                        new ComponentRegistration(
                            Guid.NewGuid(),
                            new ProvidedInstanceActivator(exported),
                            new CurrentScopeLifetime(),
                            InstanceSharing.Shared,
                            InstanceOwnership.OwnedByLifetimeScope,
                            new[] { service },
                            new Dictionary<string, object>())
                    };
                }
            }

            return new IComponentRegistration[0];
        }
    }

    public class ExportProvider : ExportDescriptorProvider
    {
        public ExportProvider()
        {
            this.Registrations = new Dictionary<Type, Func<IContainer, object>>();
        }

        public IContainer Container { get; set; }

        public IDictionary<Type, Func<IContainer, object>> Registrations { get; }

        public override IEnumerable<ExportDescriptorPromise> GetExportDescriptors(CompositionContract contract, DependencyAccessor descriptorAccessor)
        {
            if (this.Container != null && this.Registrations.TryGetValue(contract.ContractType, out var found))
            {
                yield return new ExportDescriptorPromise(
                    contract,
                    contract.ContractType.FullName,
                    true,
                    NoDependencies,
                    dependencies => ExportDescriptor.Create((context, operation) => found(this.Container), NoMetadata));
            }
        }
    }

    public interface ITest
    {
    }

    public interface ISubTest
    {
    }

    [Export(typeof(ISubTest))]
    public class SubTest1 : ISubTest
    {
    }

    [Export(typeof(ISubTest))]
    public class SubTest2 : ISubTest
    {
    }

    [Export(typeof(ITest))]
    public class Test : ITest
    {
        [ImportingConstructor]
        public Test(ILogging log, [ImportMany] ISubTest[] sub)
        { }
    }

    public interface ILogging
    { }

    public class Logging : ILogging
    { }

    public interface IPluginRegistration
    {
        void Register(ContainerBuilder builder);
    }

    [Export(typeof(IPluginRegistration))]
    public class AddPlugin : IPluginRegistration
    {
        void IPluginRegistration.Register(ContainerBuilder builder)
        {
            builder.RegisterType<Logging>().As<ILogging>().Exported();
        }
    }
}
```

**Comment by alexmg** [comment]
I have ported to .NET Standard 2.0 using the `System.ComponentModel.Composition` package and released `4.1.0` to NuGet. Please test and provide feedback if you are interested in using the updated integration.
